Docs/source/usage/workflows/extend.rst

```diff
@@ -0,0 +1,229 @@
+
+Fields
+^^^^^^
+
+This example accesses the :math:`E_x(x,y,z)` field at level 0 after every time step and calculate the largest value in it.
+
+.. code-block:: python3
+
+   # Preparation: set up the simulation
+   #   sim = picmi.Simulation(...)
+   #   ...
+
+
+   @callfromafterstep
+   def sum_rho_electrons():
+       warpx = sim.extension.warpx
+
+       # data access
+       E_x_mf = warpx.multifab(f"Efield_fp[x][l=0]")
+       E_x_np = E_x_mf.to_numpy(copy=True)  # list of arrays
+
+       # compute
+       max_E_x = max([np.max(box) for box in E_x_np])
+
+       print(f"On this process, the maximum value in E_x is {max_E_x}")
+
+
+   sim.step(nsteps=100)
+
+For further details on how to `access GPU data <https://pyamrex.readthedocs.io/en/latest/usage/zerocopy.html>`__ or compute on ``E_x``, please see the `pyAMReX documentation <https://pyamrex.readthedocs.io/en/latest/usage/compute.html#fields>`__.
+
+
+High-Level Field Wrapper
+""""""""""""""""""""""""
+
+.. note::
+
+   TODO
+
+.. note::
+
+   TODO: What are the benefits of using the high-level wrapper?
+   TODO: What are the limitations (e.g., in memory usage or compute scalability) of using the high-level wrapper?
+
+
+Particles
+^^^^^^^^^
+
+.. code-block:: python3
+
+   # Preparation: set up the simulation
+   #   sim = picmi.Simulation(...)
+   #   ...
+
+   @callfromafterstep
+   def my_after_step_callback():
+       warpx = sim.extension.warpx
+
+       # data access
+       multi_pc = libwarpx.warpx.multi_particle_container()
+       pc = mypc.get_particle_container_from_name("electrons")
+
+       # compute
+       for lvl in range(pc.finest_level + 1):
+           # get every local chunk of particles
+           for pti in pc.iterator(pc, level=lvl):
+               # additional compile-time and runtime attributes in SoA format
+               soa = pti.soa().to_numpy()
+
+               # write to all particles in the chunk
+               for soa_real in soa.real:
+                   soa_real[()] = 42.0
+
+               for soa_int in soa.int:
+                   soa_int[()] = 12
+
+
+   sim.step(nsteps=100)
+
+For further details on how to `access GPU data <https://pyamrex.readthedocs.io/en/latest/usage/zerocopy.html>`__ or compute on ``electrons``, please see the `pyAMReX documentation <https://pyamrex.readthedocs.io/en/latest/usage/compute.html#particles>`__.
+
+
+High-Level Particle Wrapper
+"""""""""""""""""""""""""""
+
+.. note::
+
+   TODO: What are the benefits of using the high-level wrapper?
+   TODO: What are the limitations (e.g., in memory usage or compute scalability) of using the high-level wrapper?
+
+Particles can be added to the simulation at specific positions and with specific attribute values:
+
+.. code-block:: python
+
+   from pywarpx import particle_containers, picmi
+
+   # ...
+
+   electron_wrapper = particle_containers.ParticleContainerWrapper("electrons")
+
+
```

Comment:
This is missing the actual example of adding particles:
```
electron_wrapper.add_particles(x=x, y=y, z=z, ux=ux, uy=uy, uz=uz, w=weight)

where x, y, z, ux, uy, uz, weight are either single values or numpy arrays of particle coordinates.
```

Reply:
The
```
.. autoclass:: pywarpx.particle_containers.ParticleContainerWrapper
   :members:
```
should contain the method:
![image](https://github.com/ECP-WarpX/WarpX/assets/1353258/f9ffd41b-7df8-4bcf-b1d4-1eb081e30437)

Reply:
Ah, okay. I thought that was perhaps the intention but wasn't sure if you also meant to have a code snippet that included it. Thanks for clarifying!

Reply:
Yes, if we want a code snippet we should add it as an example in that doc string :)